### PR TITLE
[Fix] DDP duplicate logs handling

### PR DESCRIPTION
**Pull request by sleep3r**
## Motivation

Essentially, PyTorch DDP attaches StreamHandler directly after first forward() and by default logger.root.handlers may be empty list

## Modification

Added manual append to root.handlers list of StreamHandler with level ERROR in case of empty list

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by zhouzaida** [comment]
Hi @sleep3r ,thanks for your contribution. What is your PyTorch version?

**Comment by sleep3r** [comment]
@zhouzaida 
Tested with **PyTorch 1.10.1**

I tried your logging utils with my trainloop and figured out some strange logging behaviour, let me show you some example.

**Comment by sleep3r** [comment]
Снимок экрана 2022-01-23 в 14 59 03

**Comment by sleep3r** [comment]
Here you could see, that by default the logger has not any handlers and DDP attaches one after forward of batch with idx 1. Therefore, the following logs are duplicated

**Comment by sleep3r** [comment]
I may be missing something in case of influence on downstream or other related projects, but it seems for me to be a good notice for someone with a possible similar problem

**Comment by zhouzaida** [comment]
Thank you very much for the description. BTW, is it convenient for you to test that torch<1.10.1 has similar behavior

**Comment by sleep3r** [comment]
@zhouzaida 
I've just tested. So, seems like this behaviour starts from torch==1.9.0. 

1.8.0 and 1.8.1 are fine

**Comment by zhouzaida** [comment]
> @zhouzaida I've just tested. So, seems like this behaviour starts from torch==1.9.0.
> 
> 1.8.0 and 1.8.1 are fine

Got it.

**Comment by HAOCHENYE** [comment]
LGTM. But it seems that the propagate attribute of Logger causes the duplicate log. Can set `logger.propagate` to False solve the problem of duplicate logs?  This way we don't need the change attribute of the root logger any more。

**Comment by sleep3r** [comment]
@HAOCHENYE
Well, yes logger.propagate=False solves this issue

**Comment by sleep3r** [comment]
But this fix doesn't inherit the stated one in utils/logging.py  ¯\_(ツ)_/¯

**Comment by sleep3r** [comment]
I mean, is it legit to replace all those StreamHandler setting stuff just by logger.propagate = False?

**Comment by HAOCHENYE** [comment]
The `init_process_group` function in Pytorch 1.8 adds a `StreamHandler` to the root logger, so if `get_logger` is called after `init_process_group`, the log level of the root logger will be `ERROR` and there will be no duplicate printing. Pytorch 1.10 and Pytorch1.7 do not attach a StreamHandler to the root logger during the init_process_group phase but during the forward phase of the DDP (`Logging.info` is called, which adds the default `StreamHandler`(`INFO`) to the root logger. `Logging.info `is also called in Pytorch1.8, while the log level of `StreamHandler` was set to `ERROR` by `get_logger`). The logging level of the `StreamHandler` at this point is `INFO`, so there will be a duplicate log caused by the propagate mechanism. Since the duplicate log is caused by the propagate property of the child logger, should we set the property to False? Here is the logger setting for Detectron2 https://github.com/facebookresearch/detectron2/blob/cbbc1ce26473cb2a5cc8f58e8ada9ae14cb41052/detectron2/ utils/logger.py

Translated with www.DeepL.com/Translator (free version)

**Comment by HAOCHENYE** [comment]
LGTM. I‘ll do further testing in the downstream codebase.

**Review by HAOCHENYE — APPROVED** [review]

File: mmcv/utils/logging.py [hunk 1]
```diff
@@ -38,16 +39,7 @@ def get_logger(name, log_file=None, log_level=logging.INFO, file_mode='w'):
...
-    # Starting in 1.8.0, PyTorch DDP attaches a StreamHandler <stderr> (NOTSET)
-    # to the root logger. As logger.propagate is True by default, this root
-    # level handler causes logging messages from rank>0 processes to
-    # unexpectedly show up on the console, creating much unwanted clutter.
-    # To fix this issue, we set the root logger's StreamHandler, if any, to log
-    # at the ERROR level.
-    for handler in logger.root.handlers:
-        if type(handler) is logging.StreamHandler:
-            handler.setLevel(logging.ERROR)
+    logger.propagate = False
```
Comment (zhouzaida) [inline mmcv/utils/logging.py:42]: We need to add a comment.

Comment (sleep3r) [reply mmcv/utils/logging.py:42]: @zhouzaida 
Please, check is it will be ok comment before 42th string:

# Handling of duplicate console logs.
# PyTorch DDP attaches a StreamHandler <stderr> (NOTSET)
# to the root logger. As logger.propagate is True by default, this root
# level handler causes logging messages from rank>0 processes to
# unexpectedly show up on the console, creating much unwanted clutter.

Comment (zhouzaida) [reply mmcv/utils/logging.py:42]: Append the sentence 'More details can be found at PR #1683.'

Comment (HAOCHENYE) [reply mmcv/utils/logging.py:42]: Hi @sleep3r , is there any progress?

**Comment by teamwong111** [comment]
Hi @sleep3r , is there any progress?

File: mmcv/utils/logging.py (hunk 1 above)
Comment (sleep3r) [reply mmcv/utils/logging.py:48]: @HAOCHENYE @zhouzaida 
Sorry guys for this delay. I was shocked a bit due to Russian invasion of Ukraine :(

I've updated the comment.

Comment (sleep3r) [reply mmcv/utils/logging.py:48]: But now, tests aren't passing...

Снимок экрана 2022-05-11 в 23 53 36

**Comment by zhouzaida** [comment]
Hi @sleep3r , tests failed. https://app.circleci.com/pipelines/github/open-mmlab/mmcv/582/workflows/fe9adf11-579c-45c9-8802-4a15f74bb9d9/jobs/1207?invite=true#step-110-124

**Comment by sleep3r** [comment]
@zhouzaida 
Yep, I mentioned this failed test above. But I haven't any thoughts how to fix it, unfortunately. Essentially, my new commit contains only some comment changes

**Comment by sleep3r** [comment]
I could try the former fix without .propagate

**Comment by sleep3r** [comment]
Lucky, it works

**Comment by HAOCHENYE** [comment]
@sleep3r thanks for your reply~.
If we set default StreamHandler for `logging.root` in `get_logger`, It seems there will be no message shown in the terminal by calling logging.warning('xxx'):
```python
import logging
sh = logging.StreamHandler()
sh.setLevel(logging.ERROR)
logging.root.addHandler(sh)
``` 
Is this behavior reasonable? Besides, what will happen if we set `propagate=False`, maybe there is some corner case I've not concerned.

**Comment by sleep3r** [comment]
@HAOCHENYE 
I can't reproduce the behaviour with warning logging. Setting `propagate=False` leads to fallen tests, and I don't have much about this question too.

**Comment by HAOCHENYE** [comment]
> @HAOCHENYE I can't reproduce the behaviour with warning logging. Setting `propagate=False` leads to fallen tests, and I don't have much about this question too.



> @HAOCHENYE I can't reproduce the behaviour with warning logging. Setting `propagate=False` leads to fallen tests, and I don't have much about this question too.

Hi~ sorry for my late reply. This is the original issue
https://github.com/open-mmlab/mmcv/issues/1691 
I think it could be better if we do not need to change the root logger

**Comment by OpenMMLab-Assistant003** [comment]
Hi @sleep3r！We are grateful for your efforts in helping improve this open-source project during your personal time. 
Welcome to join OpenMMLab Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA
If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤ @sleep3r
